File: companies/TikTok/74.search-a-2-d-matrix.java
/**
 * Algorithm:
 * Treat the matrix as one sorted array of length m * n. Binary-search flattened
 * indices and map mid to matrix[mid / n][mid % n].
 *
 * Complexity:
 * Time O(log(mn)), space O(1).
 */
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        int left = 0;
        int right = m * n - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int value = matrix[mid / n][mid % n];
            if (value == target) {
                return true;
            }
            if (value < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return false;
    }
}

File: companies/AMD/74.search-a-2-d-matrix.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        /*
        Approach:
        The matrix is globally sorted if viewed as a flattened array. Binary
        search over indices [0, m*n) and map mid to row mid/n and column mid%n.

        Complexity: O(log(m*n)) time and O(1) space.
        */
        int m = (int)matrix.size(), n = (int)matrix[0].size();
        int left = 0, right = m * n - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int value = matrix[mid / n][mid % n];
            if (value == target) return true;
            if (value < target) left = mid + 1;
            else right = mid - 1;
        }
        return false;
    }
};
